### ISSUE TITLE
Performance improvements

### ISSUE THREAD

**Issue by peterlarkin** [niceandserious/crafty-vagrant#2]
I think it'd be great if we can spend some time looking into the speed of crafty. It feels like a real drag sometimes.

I understand that Craft in dev mode is always going to be bit a bit slower but I wonder if there's anything we can do to speed it up, possibly on the Vagrant box?

**Comment by balazscsaba2006** [comment]
@nickautomatic Actually even on Windowa it will work, with a help of a plugin: https://github.com/winnfsd/vagrant-winnfsd

Happy coding!

**Comment by nickautomatic** [comment]
@balazscsaba2006 Thanks for that - looks interesting! Have you tried it with Vagrant on Windows at all? If so, did you find it makes a noticeable difference to performance?

**Comment by balazscsaba2006** [comment]
Yes, we are using it for almost a year now on the current project. The performance boost is more visible with heavier systems, eg. based on Symfony, and the speedup is a couple 100%s. I really recommend it for Windows users.

**Comment by nickautomatic** [comment]
Ah, that sounds really promising! I'll definitely give this a try when I have a moment. Thanks again for the suggestion.

**Comment by nickautomatic** [comment]
Another possible performance improvement (jotting here while it's fresh in my mind) - Twig has a PHP extension that improves performance of compiling Twig templates. Adding that to the Vagrant box might give another speed boost.

**Comment by balazscsaba2006** [comment]
Exactly, [this one](http://twig.sensiolabs.org/doc/installation.html)
I'm using it on production, but don't expect a huge improvement, it will be around 10-15%, but it's worth it.

**Comment by nickautomatic** [comment]
Ah, interesting - thanks again @balazscsaba2006. Good to know it makes a noticeable difference! Now I just need to find some time to get it set up...

**Comment by balazscsaba2006** [comment]
@nickautomatic This is the best part, I've installed it on 4 servers in 10 minutes, so it is just a plug and play really.

**Comment by nickautomatic** [comment]
@balazscsaba2006 Excellent, that sounds promising. I still have to write a Puppet script to get it to install in Crafty Vagrant, but sounds like it should be easy enough in that case.

**Comment by nickautomatic** [comment]
I've just seen that [Ubuntu 16.04 uses PHP 7](https://www.symfony.fi/entry/php-7-included-in-ubuntu-16-04-lts-xenial-xerus) as its default PHP version. Since that's meant to have major performance improvements over PHP 5, I'd suggest we bump Crafty up to Ubuntu 16.04 in the near future...

**Comment by nickautomatic** [comment]
An Ubuntu 16.04 / PHP 7 / Apache 2.4 / MySQL 5.7 environment is now available on a feature branch ([feature/php7.0](https://github.com/niceandserious/crafty-vagrant/tree/feature/php7.0))

This should be significantly more performant, but would be good to test this a bit before merging it into master. If anyone tries and has any feedback (eg. Did it improve performance? Did you run into any problems?) please do post it here!

**Comment by certainlyakey** [comment]
> Also, in the short term, might be worth setting type: "nfs" on the shared folder.

After inserting `:nfs => true` in the end of my `Vagrantfile` (removed `:owner => "www-data", group: "www-data"` as console asked), Craft gave me this message:

````
Could not find your craft/ folder. Please ensure that $craftPath is set correctly in /vagrant/app/public/index.php
````

The whole line in the config now looks like this:

````
config.vm.synced_folder "app/", "/vagrant/app/", :nfs => true
````

Why would this happen? The path hasn't changed.

**Comment by nickautomatic** [comment]
@certainlyakey It turns out that switching to NFS changes the way permissions are handled, which I suspect is what's causing the problem you're experiencing here. I only suggested it because I thought it might be a "quick win", improving performance a bit on Unix-like host OSes without causing problems on Windows. But since it introduces problems of its own I don't think it's worthwhile (for us, anyway - Crafty Vagrant is intended to work on OSX, Linux, or Windows). Switching to PHP7 + Apache 2.4 - which you can try on the php7.0 feature branch - seemed like a better bet, since it provides significant performance improvements regardless of host OS. In future we might revisit this, but we'd probably more likely go with rsync than NFS for Vagrant's synced folders I think (since that potentially works on all host OSes).

If you do get anywhere with NFS, do let us know (I'd be interested to see what you did to get it working), but for the above reasons there are no plans to switch to NFS with Crafty Vagrant, for the moment, anyway. If you'd like better performance, do try the [feature/php7.0](https://github.com/niceandserious/crafty-vagrant/tree/feature/php7.0) branch (which we're still testing) and let us know how you get on.

**Comment by certainlyakey** [comment]
Thanks for your reply! I'm already on the php7 - even though i installed it myself to a machine based on `master`, so Apache version is probably is not that new. The performance is better by (subjectively) somewhat 20-30%, so it's bearable though still leaves a lot to be desired.

If I'll try again to get somewhere with NFS, I'll give some instructions here.

**Comment by j-greig** [comment]
Any update on the possibility of NFS compatibility?

**Comment by nickautomatic** [comment]
@j-greig My previous comment still holds: we have no plans to switch to NFS, I'm afraid. Please do feel free to send a pull request though if you (or anyone else) do manage to get it working!
